File: Backend_Extras-Miller/White_auto.py
import cv2
import numpy as np
from typing import List


def order_points(pts: np.ndarray) -> np.ndarray:
    """
    Order 4 points in the order: TL, TR, BR, BL.
    pts: (4,2) array of float32/float64
    returns: (4,2) float32 array ordered as TL, TR, BR, BL
    """
    pts = np.asarray(pts, dtype=np.float32)

    # sums and diffs
    s = pts[:, 0] + pts[:, 1]          # x + y
    d = pts[:, 1] - pts[:, 0]          # y - x

    tl_idx = int(np.argmin(s)) # Smallest sum (TL)
    br_idx = int(np.argmax(s)) # largest sum (BR)

    # remaining two indices
    all_idx = np.array([0, 1, 2, 3])
    remain = all_idx[(all_idx != tl_idx) & (all_idx != br_idx)]

    # smaller (y - x) is TR, other is BL
    if d[remain[0]] < d[remain[1]]:
        tr_idx = int(remain[0])
        bl_idx = int(remain[1])
    else:
        tr_idx = int(remain[1])
        bl_idx = int(remain[0])

    ordered = np.array(
        [pts[tl_idx], pts[tr_idx], pts[br_idx], pts[bl_idx]],
        dtype=np.float32
    )
    return ordered


"""Example:
     
           (0, 0) -----------------------   (1919, 0)
             |                                  |
             |                                  |
             |                                  |
        (0, 1079) ----------------------- (1919, 1079)

         (349, 199) ------------------------ (2545, 409)
             |                                  |
             |                                  |
             |                                  |
        (444, 1475) ------------------------- (2525, 1437)
*/
"""

# Detect the 4 corners of the white image through bright contrast:
def detect_white_screen_corners(
    img_bgr: np.ndarray,
    adaptive_block_size: int = 75,
    adaptive_C: int = -10,
    blur_ksize: int = 51,
    diff_thresh: int = 20,
    close_ksize: int = 25,
    approx_eps_frac: float = 0.02
) -> List[List[float]]:
    
    # If no 4 point screen contour is detected:
    if img_bgr is None or img_bgr.size == 0:
        raise ValueError("Empty/invalid image input")

    # STEP 1 — Grayscale (Removes color completely and keeps only brightness) (RGB -> grayscale)
    gray = cv2.cvtColor(img_bgr, cv2.COLOR_BGR2GRAY)

    # STEP 2 — Adaptive threshold (bright regions)
    # block size must be odd and >= 3
    # Odd to detect the nearby pixels from the center when computing the threshold
    if adaptive_block_size < 3:
        adaptive_block_size = 3
    if adaptive_block_size % 2 == 0:
        adaptive_block_size += 1

    thresh = cv2.adaptiveThreshold(
        gray, 255,
        cv2.ADAPTIVE_THRESH_GAUSSIAN_C,
        cv2.THRESH_BINARY,
        adaptive_block_size,
        adaptive_C
    )

    # STEP 3 — Low variance mask (uniform areas) Finds the smooth/constant brightness:
    # blur kernel must be odd and >= 3
    if blur_ksize < 3:
        blur_ksize = 3
    if blur_ksize % 2 == 0:
        blur_ksize += 1

    # Smooth image with Gaussian blur so uniform regions remain similar while noise is reduced
    blur_img = cv2.GaussianBlur(gray, (blur_ksize, blur_ksize), 0) 

    # Gets the difference to find the smooth vs textured regions of the white image
    diff = cv2.absdiff(gray, blur_img) 

    # Used to keep smooth areas
    _, diff = cv2.threshold(diff, diff_thresh, 255, cv2.THRESH_BINARY_INV) # returns threshold value used and resulting thresholded image

    # Combine brightness + uniformity mask
    # white = bright + smooth regions (binary mask 255 (1 for ususal cases outside of OpenCV))
    # black = everything else (binary mask 0)
    combined = cv2.bitwise_and(thresh, diff)

    # STEP 4 — Morph close (fill holes)
    kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (close_ksize, close_ksize))
    combined = cv2.morphologyEx(combined, cv2.MORPH_CLOSE, kernel)

    # STEP 5 — Find contours and choose biggest 4-point polygon
    contours, _ = cv2.findContours(combined, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

    best_quad = None
    max_area = 0.0

    for c in contours:
        perimeter = cv2.arcLength(c, True)
        approx = cv2.approxPolyDP(c, approx_eps_frac * perimeter, True) # reduces points in the ocntur

        if len(approx) == 4: # Finds corners and area inside the shape
            area = cv2.contourArea(approx)
            if area > max_area:
                max_area = area
                best_quad = approx

    if best_quad is None:
        raise ValueError("No screen detected (no 4-point contour found).")

    pts = best_quad.reshape(4, 2).astype(np.float32) # gets the points from the max white contrast area
    ordered = order_points(pts)

    return ordered.tolist()


def warp_to_1920x1080(img_bgr: np.ndarray, corners_tl_tr_br_bl: List[List[float]]) -> np.ndarray:
    """
    Warp the input image into a 1920x1080 rectangle using TL,TR,BR,BL corners.
    """
    src = np.array(corners_tl_tr_br_bl, dtype=np.float32)
    dst = np.array([[0, 0], [1919, 0], [1919, 1079], [0, 1079]], dtype=np.float32)

    H = cv2.getPerspectiveTransform(src, dst)
    warped = cv2.warpPerspective(img_bgr, H, (1920, 1080))
    return warped

# Draw the red circles on the detected corners
def draw_corners(img_bgr: np.ndarray, corners_tl_tr_br_bl: List[List[float]], radius: int = 20) -> np.ndarray:
    out = img_bgr.copy()
    for x, y in corners_tl_tr_br_bl:
        cv2.circle(out, (int(x), int(y)), radius, (0, 0, 255), -1)
    return out

File: Backend_Extras-Miller/White_auto.cpp
#include <opencv2/opencv.hpp>
using namespace cv;

// ORDER 4 POINTS (TL, TR, BR, BL) with OpenCV Point2f class
std::vector<Point2f> orderPoints(const std::vector<Point2f>& pts) { 
    std::vector<Point2f> ordered(4);

    // compute sums and diffs (lambda functions)
    auto sum  = [](Point2f p){ return p.x + p.y; }; // (x + y)
    auto diff = [](Point2f p){ return p.y - p.x; }; // (y - x)

    /* 
    General logic:

    TL = smallest sum  (x + y)
    TR = smallest diff (y - x)
    BR = largest sum   (x + y)
    BL = largest diff  (y - x)

    */

// Example:
 /*.     
         (349, 199) ------------------------ (2545, 409)
             |                                  |
             |                                  |
             |                                  |
        (444, 1475) ------------------------- (2525, 1437)
*/

    // First Identify TL & BR corners:
    int tl = 0, br = 0;
    for (int i = 1; i < 4; i++) {
        if (sum(pts[i]) < sum(pts[tl])) tl = i;
        if (sum(pts[i]) > sum(pts[br])) br = i;
    }
  
    // Elminiate TL and BR and find the remaining TR and BL:
    std::vector<int> remain;
    for (int i = 0; i < 4; i++)
        if (i != tl && i != br) remain.push_back(i);

    // Distinguish TR vs BL using (y - x) --- point with smaller (y - x) is TR
    int tr = diff(pts[remain[0]]) < diff(pts[remain[1]]) ? remain[0] : remain[1];
    int bl = tr == remain[0] ? remain[1] : remain[0];

    // Return ordered points:
    ordered[0] = pts[tl]; // getPerspectiveTransform expects this order
    ordered[1] = pts[tr];
    ordered[2] = pts[br];
    ordered[3] = pts[bl];

    return ordered;
}


int main() {

    Mat img = imread("new_white_img.jpg");
    if (img.empty()) {
        std::cout << "FAILED TO LOAD new_white_img.jpg\n";
        return -1;
    }

    // STEP 1 — Grayscale (Removes color completely and keeps only brightness)
    Mat gray;
    cvtColor(img, gray, COLOR_BGR2GRAY);

    // STEP 2 — Adaptive threshold (bright regions)
    Mat thresh;
    adaptiveThreshold(gray, thresh, 255,
                      ADAPTIVE_THRESH_GAUSSIAN_C,
                      THRESH_BINARY,
                      75, -10);

    // STEP 3 — Low variance mask (uniform areas). Finds the smooth/constant brightness:
    Mat blurImg; 
    GaussianBlur(gray, blurImg, Size(51, 51), 0); // Removes noise and imperfections by replacing it with the average 51x51 neighborhood around it


    // Isolate the smooth, uniform screen region:
    Mat diff; 
    absdiff(gray, blurImg, diff); // Compute the absolute difference between the gray and blur (positive value)
    threshold(diff, diff, 20, 255, THRESH_BINARY_INV); // Within Threshold of 20 if < 20 --> White ELSE: BLACK

    // Combine brightness + uniformity mask
    Mat combined;
    bitwise_and(thresh, diff, combined);

    // STEP 4 — Morph close (Makes it a solid filled in white shape)
    Mat kernel = getStructuringElement(MORPH_RECT, Size(25, 25)); 
    morphologyEx(combined, combined, MORPH_CLOSE, kernel); // Smooth the inside of the mask by filling in the holes

    // STEP 5 — Find contours (the outline around the shape)
    std::vector<std::vector<Point>> contours;
    findContours(combined, contours, RETR_EXTERNAL, CHAIN_APPROX_SIMPLE);

    double maxArea = 0;
    std::vector<Point> bestQuad;

    for (auto &c : contours) {
        double perimeter = arcLength(c, true);
        std::vector<Point> approx;
        approxPolyDP(c, approx, 0.02 * perimeter, true); // Keeps the corners (derivations from the straight line segments) 
        // Big geometric changes (segment to corner to segment) are kept. Small changes like edge noises are removed.

        if (approx.size() == 4) {
            double area = contourArea(approx);
            if (area > maxArea) {
                maxArea = area;
                bestQuad = approx;
            }
        }
    }

    if (bestQuad.empty()) {
        std::cout << "No screen detected.\n";
        return -1;
    }

    // Convert to float points (from Point to Point2f) for warpPerspective
    std::vector<Point2f> pts;
    for (auto &p : bestQuad)
        pts.push_back(Point2f(p.x, p.y));

    // ORDER POINTS: TL, TR, BR, BL
    std::vector<Point2f> ordered = orderPoints(pts);

    // Print ordered corners:
    std::cout << "\n===== DETECTED CORNER COORDINATES =====\n";
    std::cout << "TL: (" << ordered[0].x << ", " << ordered[0].y << ")\n";
    std::cout << "TR: (" << ordered[1].x << ", " << ordered[1].y << ")\n";
    std::cout << "BR: (" << ordered[2].x << ", " << ordered[2].y << ")\n";
    std::cout << "BL: (" << ordered[3].x << ", " << ordered[3].y << ")\n";
    std::cout << "========================================\n\n";

    // STEP 6 — Draw detected corners
    Mat debug = img.clone();
    for (auto &p : ordered)
        circle(debug, p, 20, Scalar(0, 0, 255), -1);

    imshow("Detected Corners", debug);
    waitKey(0);

    // STEP 7 — Warp to 1920×1080
    std::vector<Point2f> dstPts = {
        Point2f(0, 0), Point2f(1919, 0),
        Point2f(1919, 1079), Point2f(0, 1079)
    };

    Mat H = getPerspectiveTransform(ordered, dstPts);
    Mat warped;
    warpPerspective(img, warped, H, Size(1920, 1080));

    imshow("Warped 1920x1080", warped);
    waitKey(0);

    return 0;
}


// Compile: g++ -std=c++17 white_auto.cpp -o white_auto `pkg-config --cflags --libs opencv4`
// Run: ./white_auto

